README
NAME
    Moops - Moops Object-Oriented Programming Sugar

SYNOPSIS
       use Moops;
   
       role NamedThing {
          has name => (is => "ro", isa => Str);
       }
   
       class Person with NamedThing;
   
       class Company with NamedThing;
   
       class Employee extends Person {
          has job_title => (is => "rwp", isa => Str);
          has employer  => (is => "rwp", isa => InstanceOf["Company"]);
      
          method change_job ( Object $employer, Str $title ) {
             $self->_set_job_title($title);
             $self->_set_employer($employer);
          }
      
          method promote ( Str $title ) {
             $self->_set_job_title($title);
          }
       }

STATUS
    Unstable.

    Until version 1.000, stuff might change, but not without good reason.

DESCRIPTION
    Moops is sugar for declaring and using roles and classes in Perl.

    The syntax is inspired by MooseX::Declare, and Stevan Little's
    p5-mop-redux project (which is in turn partly inspired by Perl 6).

    Moops has fewer than half of the dependencies as MooseX::Declare, loads in
    about 25% of the time, and the classes built with it run significantly
    faster. Moops does not use Devel::Declare, instead using Perl's pluggable
    keyword API; *this requires Perl 5.14 or above*.

    Moops uses Moo to build classes and roles by default, but allows you to
    use Moose if you desire. (And Mouse experimentally.)

  Classes
    The `class` keyword declares a class:

       class Foo {
          # ...
       }

    A version number can be provided:

       class Foo 1.2 {
          # ...
       }

    If no version is provided, your class' $VERSION variable is set to the
    empty string; this helps the package be seen by Class::Load.

    If your class extends an existing class through inheritance, or consumes
    one or more roles, these can also be provided when declaring the class.

       class Foo::Bar 1.2 extends Foo 1.1 with Magic::Monkeys {
          # ...
       }

    If you use Moops within a package other than `main`, then package names
    used within the declaration are "qualified" by that outer package, unless
    they contain "::". So for example:

       package Quux;
       use Moops;
   
       class Foo { }       # declares Quux::Foo
   
       class Xyzzy::Foo    # declares Xyzzy::Foo
          extends Foo { }  # ... extending Quux::Foo
   
       class ::Baz { }     # declares Baz

    If you wish to use Moose or Mouse instead of Moo; include that in the
    declaration:

       class Foo using Moose {
          # ...
       }

    It's also possible to create classes `using Tiny` (Class::Tiny), but
    there's probably little point in it, because Moops uses Moo internally, so
    the more capable Moo is already loaded and in memory.

    (The `using` option is exempt from the package qualification rules
    mentioned earlier.)

    Moops uses MooseX::MungeHas in your classes so that the `has` keyword
    supports some Moo-specific features, even when you're using Moose or
    Mouse. Specifically, it supports `is => 'rwp'`, `is => 'lazy'`, `builder
    => 1`, `clearer => 1`, `predicate => 1`, and `trigger => 1`. If you're
    using Moo, the MooX::late extension is enabled too, which allows
    Moose-isms in Moo too. With the combination of these features, there
    should be very little difference between Moo, Mouse and Moose `has`
    keywords.

    Moops uses Lexical::Accessor to provide you with private (lexical)
    attributes - that is, attributes accessed via a coderef method in a
    lexical variable.

       class Foo {
          lexical_has foo => (
             isa      => Int,
             accessor => \(my $_foo),
             default  => 0,
          );
          method increment_foo () {
             $self->$_foo( 1 + $self->$_foo );
          }
          method get_foo () {
             return $self->$_foo;
          }
       }
   
       my $x = Foo->new;
       $x->increment_foo();     # ok
       say $x->get_foo();       # says "1"
       $x->$_foo(42);           # dies; $_foo does not exist in this scope

    Moose classes are automatically accelerated using MooseX::XSAccessor if
    it's installed.

    Note that it is possible to declare a class with an empty body; use a
    trailing semicolon.

       class Employee extends Person with Employment;

    If using Moose or Mouse, classes are automatically made immutable.

    namespace::sweep is automatically used in all classes.

    Between the class declaration and its body, Attribute::Handlers-style
    attributes may be provided:

       class Person :mutable {
          # ...
       }
   
       class Employee extends Person with Employment :mutable;

    The following attributes are defined for classes:

    *   `:assertions` - enables assertion checking (see below)

    *   `:dirty` - suppresses namespace::sweep

    *   `:fp` - use Function::Parameters instead of Kavorka

    *   `:mutable` - suppresses making Moose classes immutable

    *   `:ro` - make attributes declared with `has` default to 'ro'

    *   `:rw` - make attributes declared with `has` default to 'rw'

    *   `:rwp` - make attributes declared with `has` default to 'rwp'

  Roles
    Roles can be declared similarly to classes, but using the `role` keyword.

       role Stringable
          using Moose     # we know you meant Moose::Role
       {
          # ...
       }

    Roles do not support the `extends` option.

    Roles can be declared to be `using` Moo, Moose, Mouse or Tiny. (Note that
    if you're mixing and matching role frameworks, there are limitations to
    which class builders can consume which roles. Mouse is generally the least
    compatible; Moo and Moose classes should be able to consume each others'
    roles; Moo can also consume Role::Tiny roles.)

    If roles use Moo, the MooX::late extension is enabled.

    namespace::sweep is automatically used in all roles.

    Roles take similar Attribute::Handlers-style attributes to classes, but
    don't support `:mutable`.

   A note on consuming roles
    In a standard:

       class MyClass with MyRole {
          ...;
       }

    You should note that role composition is delayed to happen at the *end* of
    the class declaration. This is usually what you want.

    However the interaction between method modifiers and roles is complex, and
    *sometimes* you'll want the role to be applied to the class part-way
    through the declaration. In this case you can use a `with` statement
    *inside* the class declaration:

       class MyClass {
          ...;
          with "MyRole";
          ...;
       }

  Namespaces
    The `namespace` keyword works as above, but declares a package without any
    class-specific or role-specific semantics.

       namespace Utils {
          # ...
       }

    namespace::sweep is not automatically used in namespaces.

    Attribute::Handlers-style attributes are supported for namespaces, but
    most of the built-in attributes make any sense without class/role
    semantics. (`:assertions` does.) Traits written as Moops extensions may
    support namespaces.

  Functions and Methods
    Moops uses Kavorka to declare functions and methods within classes and
    roles. Kavorka provides the `fun` and `method` keywords.

       class Person {
          use Scalar::Util 'refaddr';
      
          has name => (is => 'rwp');    # Moo attribute
      
          method change_name ( Str $newname ) {
             $self->_set_name( $newname )
                unless $newname eq 'Princess Consuela Banana-Hammock';
          }
      
          fun is_same_as ( Object $x, Object $y ) {
             refaddr($x) == refaddr($y)
          }
       }
   
       my $phoebe = Person->new(name => 'Phoebe');
       my $ursula = Person->new(name => 'Ursula');
   
       Person::is_same_as($phoebe, $ursula);   # false

    Note function signatures use type constraints from Types::Standard;
    MooseX::Types and MouseX::Types type constraints should also work,
    *provided you use their full names, including their package*.

    The `is_same_as` function above could have been written as a class method
    like this:

       class Person {
          # ...
          method is_same_as ( $class: Object $x, Object $y ) {
             refaddr($x) == refaddr($y)
          }
       }
   
       # ...
       Person->is_same_as($phoebe, $ursula);   # false

    The `method` keyword is not provided within packages declared using
    `namespace`; only within classes and roles.

    See also Kavorka::Manual::Methods and Kavorka::Manual::Functions.

    Within Moose classes and roles, the MooseX::KavorkaInfo module is loaded,
    to allow access to method signatures via the meta object protocol. (This
    is currently broken for `around` method modifiers.)

    In Moops prior to 0.025, Function::Parameters was used instead of Kavorka.
    If you wish to continue to use Function::Parameters in a class you can use
    the `:fp` attribute:

       class Person :fp {
          ...;
       }

    Or to do so for all classes in a lexical scope:

       use Moops function_parameters_everywhere => 1;
       class Person {
          ...;
       }

    Or the environment variable `MOOPS_FUNCTION_PARAMETERS_EVERYWHERE` can be
    set to true to enable it globally, but this feature is likely to be
    removed eventually.

  Method Modifiers
    Within classes and roles, `before`, `after` and `around` keywords are
    provided for declaring method modifiers. These use the same syntax as
    `method`.

    If your class or role is using Moose or Mouse, then you also get `augment`
    and `override` keywords.

    See also Kavorka::Manual::MethodModifiers.

  Multi Methods
    Moops uses Kavorka to implement multi subs and multi methods.

    See also Kavorka::Manual::MultiSubs.

  Type Constraints
    The Types::Standard type constraints are exported to each package declared
    using Moops. This allows the standard type constraints to be used as
    barewords.

    Type constraints can be used in attribute definitions (`isa`) and method
    signatures. Because Types::Standard is based on Type::Tiny, the same type
    constraints may be used whether you build your classes and roles with Moo,
    Moose our Mouse.

    Alternative libraries can be imported using the `types` option; a la:

       class Document types Types::XSD::Lite {
          has title => (is => 'rw', isa => NormalizedString);
       }

    Note that if an alternative type constraint library is imported, then
    Types::Standard is *not* automatically loaded, and needs to be listed
    explicitly:

       class Document types Types::Standard, Types::XSD::Lite {
          # ...
       }

    Type libraries built with Type::Library, MooseX::Types and MouseX::Types
    should all work.

    Bear in mind that type constraints from, say, a MooseX::Types library
    won't be usable in, say, Moo attribute definitions. However, it's possible
    to wrap them with Type::Tiny, and make them usable:

       class Foo types MooseX::Types::Common::Numeric using Moo {
          use Types::TypeTiny qw( to_TypeTiny );
      
          has favourite_number => (
             is  => 'rwp',
             isa => to_TypeTiny(PositiveInt)
          );
       }

  Type Libraries
    You can use the `library` keyword to declare a new type library:

       library MyTypes
          extends Types::Standard
          declares EmptyString, NonEmptyString {
      
          declare EmptyString,
             as Str,
             where { length($_) == 0 };
      
          declare NonEmptyString,
             as Str,
             where { length($_) > 0 };
       }
   
       class StringChecker types MyTypes {
          method check ( Str $foo ) {
             return "empty" if EmptyString->check($foo);
             return "non-empty" if NonEmptyString->check($foo);
             return "impossible?!";
          }
       }

    Libraries declared this way can extend existing type libraries written
    with Type::Library, MooseX::Types or MouseX::Types.

    Note that this also provides a solution to the previously mentioned
    problem of using MooseX::Types type libraries in Moo classes:

       library MyWrapper
          extends MooseX::Types::Common::Numeric;
   
       class Foo types MyWrapper using Moo {
          has favourite_number => (
             is  => 'rwp',
             isa => PositiveInt,
          );
       }

  Constants
    The useful constants `true` and `false` are imported into all declared
    packages. (Within classes and roles, namespace::sweep will later remove
    them from the symbol table, so they don't form part of your package's
    API.) These constants can help make attribute declarations more readable.

       has name => (is => 'ro', isa => Str, required => true);

    Further constants can be declared using the `define` keyword (see
    PerlX::Define):

       namespace Maths {
          define PI = 3.2;
       }

    Constants declared this way will *not* be swept away by namespace::sweep,
    and are considered part of your package's API.

  Assertions
    Declared packages can contain assertions (see PerlX::Assert). These are
    normally optimized away at compile time, but you can force them to be
    checked using the `:assertions` attribute.

       class Foo {
          assert(false);    # not checked; optimized away
       }
   
       class Bar :assertions {
          assert(false);    # checked; fails; throws exception
       }

  More Sugar
    strict and FATAL warnings are imported into all declared packages. However
    the `uninitialized`, `void`, `once` and `numeric` warning categories are
    explicitly excluded, as are any warnings categories added to Perl after
    version 5.14.

    Perl 5.14 features, including the `state` and `say` keywords, and sane
    Unicode string handling are imported into all declared packages.

    Try::Tiny is imported into all declared packages.

    Scalar::Util's `blessed` and Carp's `confess` are imported into all
    declared packages.

  Outer Sugar
    The "outer" package, where the `use Moops` statement appears also gets a
    little sugar: strict, the same warnings as "inner" packages, and Perl 5.14
    features are all switched on.

    true is loaded, so you don't need to do this at the end of your file:

       1;

  Custom Sugar
    It is possible to inject other functions into all inner packages using:

       use Moops imports => [
          'List::Util'      => [qw( first reduce )],
          'List::MoreUtils' => [qw( any all none )],
       ];

    This is by far the easiest way to extend Moops with project-specific
    extras.

    There is a shortcut for injecting strictures into all inner packages:

       use Moops -strict;

EXTENDING
    Moops is written to hopefully be fairly extensible.

  Extending Moops via imports
    The easiest way to extend Moops is to inject additional imports into the
    inner packages using the technique outlined in "Custom Sugar" above. You
    can wrap all that up in a module:

       package MoopsX::Lists;
       use base 'Moops';
       use List::Util ();
       use List::MoreUtils ();
   

Changes
Moops
=====

Created:      2013-06-30
Home page:    <https://metacpan.org/release/Moops>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Moops>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.033	2014-08-16

 [ Bug Fixes ]
 - Fix bugs introduced in 0.032 with role application via guards.

 [ Other ]
 - Updated: Take advantage of improvements in Type::Library/Type::Registry
   1.000000.

0.032	2014-08-13

 [ BACK COMPAT ]
 - Role composition is now delayed to the end of a class declaration.
   Matt S Trout++

 [ Documentation ]
 - Fix a typo in Moops::Manual::Objects101.
   Fixes RT#93553.
   Ron Savage++
   <https://rt.cpan.org/Ticket/Display.html?id=93553>

 [ Packaging ]
 - PerlX::Assert is no longer bundled with Moops; it has been split out
   into a separate distribution.

0.031	2014-03-23

 [ Documentation ]
 - Fix typo in manual.
   Fixes RT#93553.
   Ron Savage++
   <https://rt.cpan.org/Ticket/Display.html?id=93553>
 - Use the `true` constant instead of the number 1 in the cli-app.pl
   example.

 [ Test Suite ]
 - Test Moops in threaded Perls.
   Aaron James Trevena++

 [ Other ]
 - Added: The new import parameter `-strict` enables loading strictures.pm
   within Moops scopes.
 - Avoid returning 0 from `assert` statements (PerlX::Assert).
 - Use Parse::Keyword instead of Devel::Pragma in order to establish the
   currently compiling package name. Devel::Pragma has issues under
   ithreads.
   Piers Cawley++

0.030	2014-01-30

 [ Documentation ]
 - Include an example using Moops to write a command-line app.
   Maurice Mengel++
   <https://github.com/tobyink/p5-moops/pull/1>

 [ Packaging ]
 - Require at least version 0.027 of Kavorka; this resolves a conflict
   using Moops with MooX::Options.

 [ Other ]
 - Hard-code a list of warnings categories to fatalize; this means that
   warnings categories added to future versions of Perl will not magically
   become fatal in Moops.

0.029	2014-01-04

 [ Bug Fixes ]
 - Fix availability of Lexical::Accessor within class {} blocks.

 [ Documentation ]
 - Copyright 2014.

0.028	2013-12-18

 [ Documentation ]
 - Demonstrate `is override` trait in Moops::Manual::Objects101.

 [ Packaging ]
 - Indicate conflict with MooseX::Types::Common 0.001011.
 - Indicate conflict with MooseX::Types::URI 0.05.

0.027	2013-12-16

 [ Documentation ]
 - Include an example using multi methods to implement a class method and
   object method with the same name as each other.
 - Moops::Manual::Objects101.

 [ Other ]
 - Added: Private attributes via Lexical::Accessor.
 - Moose/Mouse classes/roles now have override/augment method modifier
   keywords.

0.026	2013-11-18

 - Replace Function::Parameters with Kavorka.

0.025	2013-10-09

 [ Documentation ]
 - Note that a future version of Moops may replace Function::Parameters
   with Kavorka.

 [ Other ]
 - Updated: Use Exporter::Tiny.

0.024	2013-09-27

 [ Test Suite ]
 - Keyword::Simple 0.02 has been released; no longer need to skip
   t/03lines.t.

0.023	2013-09-26

 [ Bug Fixes ]
 - Much improved (though not completely perfect) line number accounting.

 [ Documentation ]
 - Bundle an example of using Moops with Method::Signatures.

0.022	2013-09-16

 - Minor updates to work with Function::Parameters 1.0301.

0.021	2013-09-12

 - Allow version numbers to be specified for the `with`, `extends` and
   `types` options.

0.020	2013-09-11

 [ BACK COMPAT ]
 - Moops->import now takes a hash of options (including the `imports`
   option) rather than an arrayref of modules to import.

 [ Packaging ]
 - List Moose/Mouse/Moo/Class::Tiny dependencies as optional_features in
   META.json.

 [ Other ]
 - Improve Moops' extensibility via parser traits.

0.019	2013-08-30

 - Removed: Removed Class::Tiny::Antlers; this is now a separate CPAN
   distribution.

0.018	2013-08-27

 - Added: Add a `library` keyword for declaring type libraries.
 - Declared packages now get an implied BEGIN {...} block around
   themselves.

0.017	2013-08-21

 - Updated: Support Class::Tiny 0.004.

0.016	2013-08-21

 - Added: Provide a `types` option for loading type constraint libraries
   into classes, roles and namespaces.

0.015	2013-08-21

 [ Bug Fixes ]
 - Fix a reference to Moops::DefineKeyword which will only work if you
   happen to have installed Moops over the top of a pre-0.012 version of
   Moops.
 - Load feature.pm so that it's consistently exported to the outer scope.

0.014	2013-08-21

 [ Documentation ]
 - Document Class::Tiny::Antlers.

 [ Other ]
 - Added: MooseX::FunctionParametersInfo
 - Class::Tiny::Antlers now supports has \@attrs like Moose.

0.013	2013-08-20

 [ Bug Fixes ]
 - Fix test that uses Role::Tiny and Class::Tiny without declaring them
   (Test::Requires).

0.012	2013-08-20

 [ Documentation ]
 - Various documentation improvements.

 [ Other ]
 - Added: PerlX::Assert
 - Rename Moops::DefineKeyword -> PerlX::Define.

0.011	2013-08-20

 [ Documentation ]
 - Document Attribute::Handlers-style attributes as an extensibility
   mechanism.

 [ Other ]
 - Added: Class::Tiny::Antlers
 - Added: Moops::TraitFor::Keyword::dirty
 - Added: Moops::TraitFor::Keyword::mutable
 - Added: Moops::TraitFor::Keyword::ro
 - Added: Moops::TraitFor::Keyword::rw
 - Added: Moops::TraitFor::Keyword::rwp
 - Added: Support classes built `using Class::Tiny`.
 - Moose classes will now `use Moose::XSAccessor` if possible.
 - Use MooseX::MungeHas to 0.002 smooth over more differences between Moo,
   Mouse and Moose.

0.010	2013-08-19

 - Added: Parse Attribute::Handlers-style attributes attached to package
   declarations; treat these as traits for the code generator.
 - Much refactoring.
 - Rename Moops::CodeGenerator -> Moops::Keyword.

0.009	2013-08-19

 [ Bug Fixes ]
 - Fix at-runtime hook (used for method modifiers).

 [ Packaging ]
 - The test suite is now in a reasonable state.

0.008	2013-08-18

 [ Bug Fixes ]
 - Fix custom imports feature.
 - Found a remaining hard-coded list of keywords that was breaking

Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"          => "Moops Object-Oriented Programming Sugar",
  "author"            => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>", "unknown"],
  "dynamic_config"    => 1,
  "generated_by"      => "Dist::Inkt::Profile::TOBYINK version 0.016, CPAN::Meta::Converter version 2.140640",
  "keywords"          => [],
  "license"           => ["perl_5"],
  "meta-spec"         => {
                           url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                           version => 2,
                         },
  "name"              => "Moops",
  "no_index"          => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "optional_features" => {
                           Moo   => {
                                      description => "allow classes and roles to be built with Moo",
                                      prereqs     => { runtime => { suggests => { "MooX::HandlesVia" => 0 } } },
                                      x_default   => 1,
                                    },
                           Moose => {
                                      description => "allow classes and roles to be built with Moose",
                                      prereqs     => {
                                                       runtime => {
                                                         recommends => { "MooseX::XSAccessor" => 0 },
                                                         requires   => { Moose => "2.0600" },
                                                       },
                                                       test => { suggests => { "MooseX::Types::Common::Numeric" => 0 } },
                                                     },
                                      x_default   => 0,
                                    },
                           Mouse => {
                                      description => "allow classes and roles to be built with Mouse",
                                      prereqs     => { runtime => { requires => { Mouse => "1.00" } } },
                                      x_default   => 0,
                                    },
                           Tiny  => {
                                      description => "allow classes and roles to be built with Class::Tiny/Role::Tiny",
                                      prereqs     => {
                                                       runtime => {
                                                         requires => { "Class::Tiny::Antlers" => 0, "Role::Tiny" => "1.000000" },
                                                       },
                                                     },
                                      x_default   => 0,
                                    },
                         },
  "prereqs"           => {
                           configure => {
                                          requires => {
                                            "CPAN::Meta::Requirements" => "2.000",
                                            "ExtUtils::MakeMaker"      => 6.17,
                                          },
                                        },
                           develop   => { recommends => { "Dist::Inkt" => 0 } },
                           runtime   => {
                                          conflicts  => {
                                                          "MooseX::Types::Common" => "== 0.001011",
                                                          "MooseX::Types::URI"    => "== 0.04",
                                                        },
                                          recommends => { "Keyword::Simple" => 0.02 },
                                          requires   => {
                                                          "Exporter::Tiny"    => 0.026,
                                                          "Import::Into"      => "1.000000",
                                                          "Kavorka"           => 0.027,
                                                          "Keyword::Simple"   => 0.01,
                                                          "Lexical::Accessor" => 0.003,
                                                          "Module::Runtime"   => 0.013,
                                                          "Moo"               => "1.003000",
                                                          "MooseX::MungeHas"  => 0.002,
                                                          "MooX::late"        => 0.014,
                                                          "namespace::sweep"  => 0.006,
                                                          "Parse::Keyword"    => 0.006,
                                                          "perl"              => 5.014,
                                                          "PerlX::Assert"     => 0.902,
                                                          "Scalar::Util"      => 1.24,
                                                          "Scope::Guard"      => 0,
                                                          "Sub::Name"         => 0,
                                                          "true"              => 0.18,
                                                          "Try::Tiny"         => 0.12,
                                                          "Type::Utils"       => "1.000000",
                                                        },
                                        },
                           test      => {
                                          recommends => { "Test::Warnings" => 0 },
                                          requires   => { "Test::Fatal" => 0, "Test::More" => 0.96, "Test::Requires" => 0 },
                                          suggests   => { "Types::XSD::Lite" => 0.003 },
                                        },
                         },
  "provides"          => {
                           "Moops"                                                => { file => "lib/Moops.pm", version => 0.033 },
                           "Moops::ImportSet"                                     => { file => "lib/Moops/ImportSet.pm", version => 0.033 },
                           "Moops::Keyword"                                       => { file => "lib/Moops/Keyword.pm", version => 0.033 },
                           "Moops::Keyword::Class"                                => { file => "lib/Moops/Keyword/Class.pm", version => 0.033 },
                           "Moops::Keyword::Library"                              => { file => "lib/Moops/Keyword/Library.pm", version => 0.033 },
                           "Moops::Keyword::Role"                                 => { file => "lib/Moops/Keyword/Role.pm", version => 0.033 },
                           "Moops::MethodModifiers"                               => { file => "lib/Moops/MethodModifiers.pm", version => 0.033 },
                           "Moops::Parser"                                        => { file => "lib/Moops/Parser.pm", version => 0.033 },
                           "Moops::TraitFor::Keyword::assertions"                 => {
                                                                                       file => "lib/Moops/TraitFor/Keyword/assertions.pm",
                                                                                       version => 0.033,
                                                                                     },
                           "Moops::TraitFor::Keyword::dirty"                      => { file => "lib/Moops/TraitFor/Keyword/dirty.pm", version => 0.033 },
                           "Moops::TraitFor::Keyword::fp"                         => { file => "lib/Moops/TraitFor/Keyword/fp.pm", version => 0.033 },
                           "Moops::TraitFor::Keyword::mutable"                    => { file => "lib/Moops/TraitFor/Keyword/mutable.pm", version => 0.033 },
                           "Moops::TraitFor::Keyword::ro"                         => { file => "lib/Moops/TraitFor/Keyword/ro.pm", version => 0.033 },
                           "Moops::TraitFor::Keyword::rw"                         => { file => "lib/Moops/TraitFor/Keyword/rw.pm", version => 0.033 },
                           "Moops::TraitFor::Keyword::rwp"                        => { file => "lib/Moops/TraitFor/Keyword/rwp.pm", version => 0.033 },
                           "MooseX::FunctionParametersInfo"                       => { file => "lib/MooseX/FunctionParametersInfo.pm", version => 0.033 },
                           "MooseX::FunctionParametersInfo::Trait::Method"        => { file => "lib/MooseX/FunctionParametersInfo.pm", version => 0.033 },
                           "MooseX::FunctionParametersInfo::Trait::WrappedMethod" => { file => "lib/MooseX/FunctionParametersInfo.pm", version => 0.033 },
                           "PerlX::Define"                                        => { file => "lib/PerlX/Define.pm", version => 0.033 },
                         },
  "release_status"    => "stable",
  "resources"         => {
                           bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=Moops" },
                           homepage     => "https://metacpan.org/release/Moops",
                           license      => ["http://dev.perl.org/licenses/"],
                           repository   => {
                                             type => "git",
                                             url  => "git://github.com/tobyink/p5-moops.git",
                                             web  => "https://github.com/tobyink/p5-moops",
                                           },
                           X_identifier => "http://purl.org/NET/cpan-uri/dist/Moops/project",
                         },
  "version"           => 0.033,
  "x_contributors"    => [
                           "Maurice Mengel (MMAURICE) <mmaurice\@cpan.org>",
                           "Aaron James Trevena (TEEJAY) <teejay\@cpan.org>",
                         ],
};

if ($ENV{MM_INSTALL_FEATURES})
{
	my %features = %{ $meta->{optional_features} };
	my @features = sort {
		$features{$b}{x_default} <=> $features{$a}{x_default} or $a cmp $b
	} keys %features;
	
	for my $feature_id (@features)
	{
		my %feature = %{ $features{$feature_id} };
		
		next unless prompt(
			sprintf('Install %s (%s)?', $feature_id, $feature{description} || 'no description'),
			$feature{x_default} ? 'Y' : 'N',
		) =~ /^Y/i;
		
		$features{$feature_id}{x_selected} = 1;
		
		for my $stage (keys %{$feature{prereqs}})
		{
			for my $level (keys %{$feature{prereqs}{$stage}})
			{
				for my $module (keys %{$feature{prereqs}{$stage}{$level}})
				{
					$meta->{prereqs}{$stage}{$level}{$module}
						= $feature{prereqs}{$stage}{$level}{$module};
				}
			}
		}
	}
}
else
{
	print <<'MM_INSTALL_FEATURES';

** Setting the MM_INSTALL_FEATURES environment variable to true
** would allow you to choose additional features.

MM_INSTALL_FEATURES
}

my %dynamic_config;
for my $stage (keys %{$meta->{prereqs}})
{
	my $conflicts = $meta->{prereqs}{$stage}{conflicts} or next;
	eval { require CPAN::Meta::Requirements } or last;
	$conflicts = 'CPAN::Meta::Requirements'->from_string_hash($conflicts);
	
	for my $module ($conflicts->required_modules)
	{
		eval "require $module" or next;
		my $installed = eval(sprintf('$%s::VERSION', $module));
		$conflicts->accepts_module($module, $installed) or next;
		
		my $message = "\n".
			"** This version of $meta->{name} conflicts with the version of\n".
			"** module $module ($installed) you have installed.\n";
		die($message . "\n" . "Bailing out")
			if $stage eq 'build' || $stage eq 'configure';
		
		$message .= "**\n".
			"** It's strongly recommended that you update it after\n".
			"** installing this version of $meta->{name}.\n";
		warn("$message\n");
	}
}

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

